src/iterative_ensemble_smoother/experimental_sies.py

```diff
@@ -0,0 +1,418 @@
+
+
+def inversion_naive(*, W, step_length, S, C_dd, H, C_dd_cholesky=None):
+    """Naive implementation of Equation (42)."""
+    _verify_inversion_args(
+        W=W, step_length=step_length, S=S, C_dd=C_dd, H=H, C_dd_cholesky=C_dd_cholesky
+    )
+
+    # Since it's naive, we just create zeros here
+    if C_dd.ndim == 1:
+        C_dd = np.diag(C_dd)
+
+    to_invert = S @ S.T + C_dd
+    return W - step_length * (
+        W - np.linalg.multi_dot([S.T, sp.linalg.inv(to_invert), H])
+    )
+
+
+def inversion_direct(*, W, step_length, S, C_dd, H, C_dd_cholesky=None):
+    """Implementation of equation (42)."""
+    _verify_inversion_args(
+        W=W, step_length=step_length, S=S, C_dd=C_dd, H=H, C_dd_cholesky=C_dd_cholesky
+    )
+
+    # We define K as the solution to
+    # (S @ S.T + C_dd) @ K = H, so that
+    # K = (S @ S.T + C_dd)^-1 H
+    # K = solve(S @ S.T + C_dd, H)
+    if C_dd.ndim == 1:
+        # Since S has shape (m, N), and m >> N, using BLAS is typically faster:
+        # >>> S = np.random.randn(10_000, 100)
+        # >>> %timeit S @ S.T
+        # 512 ms ± 14.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+        # >>> %timeit sp.linalg.blas.dsyrk(alpha=1.0, a=S)
+        # 225 ms ± 37 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+        # Here BLAS only computes the upper triangular part of S @ S.T, but
+        # that's all we need when we call the solver with "lower=False".
+        lhs = sp.linalg.blas.dsyrk(alpha=1.0, a=S)  # Equivalent to S @ S.T
+        lhs.flat[:: lhs.shape[0] + 1] += C_dd
+    else:
+        lhs = sp.linalg.blas.dsyrk(alpha=1.0, a=S, c=C_dd, beta=1.0)  # S @ S.T + C_dd
+
+    K = sp.linalg.solve(
+        lhs,
+        H,
+        lower=False,
+        overwrite_a=True,
+        overwrite_b=False,
+        check_finite=True,
+        assume_a="pos",
+    )
+    return W - step_length * (W - np.linalg.multi_dot([S.T, K]))
+
+
+def inversion_direct_corrscale(*, W, step_length, S, C_dd, H, C_dd_cholesky=None):
+    """Implementation of equation (42), with correlation matrix scaling."""
+    _verify_inversion_args(
+        W=W, step_length=step_length, S=S, C_dd=C_dd, H=H, C_dd_cholesky=C_dd_cholesky
+    )
+
+    if C_dd.ndim == 2:
+        scale_factor = 1 / np.sqrt(np.diag(C_dd))
+        # Scale rows and columns by diagonal, creating correlation matrix R
+        R = (C_dd * scale_factor.reshape(1, -1)) * scale_factor.reshape(-1, 1)
+    else:
+        scale_factor = 1 / np.sqrt(C_dd)
+        # The correlation matrix R is simply the identity matrix in this case
+
+    # Scale rows
+    S = S * scale_factor.reshape(-1, 1)
+    H = H * scale_factor.reshape(-1, 1)
+
+    # We define K as the solution to
+    # (S @ S.T + C_dd) @ K = H, so that
+    # K = (S @ S.T + C_dd)^-1 H
+    # K = solve(S @ S.T + C_dd, H)
+    if C_dd.ndim == 1:
+        lhs = sp.linalg.blas.dsyrk(alpha=1.0, a=S)  # S.T @ S
+        lhs.flat[:: lhs.shape[0] + 1] += 1
```

Comment:
What if we use the following instead as I think it's easier to understand?

np.fill_diagonal(lhs, lhs.diagonal() + 1)